<!DOCTYPE html>
<!-- Generated by pkgdown: do not edit by hand --><html lang="en">
<head>
<meta http-equiv="Content-Type" content="text/html; charset=UTF-8">
<meta charset="utf-8">
<meta http-equiv="X-UA-Compatible" content="IE=edge">
<meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<meta name="description" content="DAISIEprep">
<title>Extending DAISIEprep ASR models • DAISIEprep</title>
<!-- favicons --><link rel="icon" type="image/png" sizes="16x16" href="../favicon-16x16.png">
<link rel="icon" type="image/png" sizes="32x32" href="../favicon-32x32.png">
<link rel="apple-touch-icon" type="image/png" sizes="180x180" href="../apple-touch-icon.png">
<link rel="apple-touch-icon" type="image/png" sizes="120x120" href="../apple-touch-icon-120x120.png">
<link rel="apple-touch-icon" type="image/png" sizes="76x76" href="../apple-touch-icon-76x76.png">
<link rel="apple-touch-icon" type="image/png" sizes="60x60" href="../apple-touch-icon-60x60.png">
<script src="../deps/jquery-3.6.0/jquery-3.6.0.min.js"></script><meta name="viewport" content="width=device-width, initial-scale=1, shrink-to-fit=no">
<link href="../deps/bootstrap-5.2.2/bootstrap.min.css" rel="stylesheet">
<script src="../deps/bootstrap-5.2.2/bootstrap.bundle.min.js"></script><!-- Font Awesome icons --><link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/all.min.css" integrity="sha256-mmgLkCYLUQbXn0B1SRqzHar6dCnv9oZFPEC1g1cwlkk=" crossorigin="anonymous">
<link rel="stylesheet" href="https://cdnjs.cloudflare.com/ajax/libs/font-awesome/5.12.1/css/v4-shims.min.css" integrity="sha256-wZjR52fzng1pJHwx4aV2AO3yyTOXrcDW7jBpJtTwVxw=" crossorigin="anonymous">
<!-- bootstrap-toc --><script src="https://cdn.jsdelivr.net/gh/afeld/bootstrap-toc@v1.0.1/dist/bootstrap-toc.min.js" integrity="sha256-4veVQbu7//Lk5TSmc7YV48MxtMy98e26cf5MrgZYnwo=" crossorigin="anonymous"></script><!-- headroom.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/headroom.min.js" integrity="sha256-AsUX4SJE1+yuDu5+mAVzJbuYNPHj/WroHuZ8Ir/CkE0=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/headroom/0.11.0/jQuery.headroom.min.js" integrity="sha256-ZX/yNShbjqsohH1k95liqY9Gd8uOiE1S4vZc+9KQ1K4=" crossorigin="anonymous"></script><!-- clipboard.js --><script src="https://cdnjs.cloudflare.com/ajax/libs/clipboard.js/2.0.6/clipboard.min.js" integrity="sha256-inc5kl9MA1hkeYUt+EC3BhlIgyp/2jDIyBLS6k3UxPI=" crossorigin="anonymous"></script><!-- search --><script src="https://cdnjs.cloudflare.com/ajax/libs/fuse.js/6.4.6/fuse.js" integrity="sha512-zv6Ywkjyktsohkbp9bb45V6tEMoWhzFzXis+LrMehmJZZSys19Yxf1dopHx7WzIKxr5tK2dVcYmaCk2uqdjF4A==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/autocomplete.js/0.38.0/autocomplete.jquery.min.js" integrity="sha512-GU9ayf+66Xx2TmpxqJpliWbT5PiGYxpaG8rfnBEk1LL8l1KGkRShhngwdXK1UgqhAzWpZHSiYPc09/NwDQIGyg==" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mark.js/8.11.1/mark.min.js" integrity="sha512-5CYOlHXGh6QpOFA/TeTylKLWfB3ftPsde7AnmhuitiTX4K5SqCLBeKro6sPS8ilsz1Q4NRx3v8Ko2IBiszzdww==" crossorigin="anonymous"></script><!-- pkgdown --><script src="../pkgdown.js"></script><meta property="og:title" content="Extending DAISIEprep ASR models">
<meta property="og:description" content="DAISIEprep">
<meta property="og:image" content="https://joshwlambert.github.io/DAISIEprep/logo.png">
<!-- mathjax --><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/MathJax.js" integrity="sha256-nvJJv9wWKEm88qvoQl9ekL2J+k/RWIsaSScxxlsrv8k=" crossorigin="anonymous"></script><script src="https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.5/config/TeX-AMS-MML_HTMLorMML.js" integrity="sha256-84DKXVJXs0/F8OTMzX4UR909+jtl4G7SPypPavF+GfA=" crossorigin="anonymous"></script><!--[if lt IE 9]>
<script src="https://oss.maxcdn.com/html5shiv/3.7.3/html5shiv.min.js"></script>
<script src="https://oss.maxcdn.com/respond/1.4.2/respond.min.js"></script>
<![endif]-->
</head>
<body>
    <a href="#main" class="visually-hidden-focusable">Skip to contents</a>
    

    <nav class="navbar fixed-top navbar-light navbar-expand-lg bg-light"><div class="container">
    
    <a class="navbar-brand me-2" href="../index.html">DAISIEprep</a>

    <small class="nav-text text-muted me-auto" data-bs-toggle="tooltip" data-bs-placement="bottom" title="">0.3.1</small>

    
    <button class="navbar-toggler" type="button" data-bs-toggle="collapse" data-bs-target="#navbar" aria-controls="navbar" aria-expanded="false" aria-label="Toggle navigation">
      <span class="navbar-toggler-icon"></span>
    </button>

    <div id="navbar" class="collapse navbar-collapse ms-3">
      <ul class="navbar-nav me-auto">
<li class="nav-item">
  <a class="nav-link" href="../reference/index.html">Reference</a>
</li>
<li class="active nav-item dropdown">
  <a href="#" class="nav-link dropdown-toggle" data-bs-toggle="dropdown" role="button" aria-expanded="false" aria-haspopup="true" id="dropdown-articles">Articles</a>
  <div class="dropdown-menu" aria-labelledby="dropdown-articles">
    <a class="dropdown-item" href="../articles/Extending_asr.html">Extending DAISIEprep ASR models</a>
    <a class="dropdown-item" href="../articles/Performance.html">Performance</a>
    <a class="dropdown-item" href="../articles/Sensitivity.html">Sensitivity</a>
    <a class="dropdown-item" href="../articles/Tutorial.html">DAISIEprep Tutorial</a>
  </div>
</li>
<li class="nav-item">
  <a class="nav-link" href="../news/index.html">Changelog</a>
</li>
      </ul>
<form class="form-inline my-2 my-lg-0" role="search">
        <input type="search" class="form-control me-sm-2" aria-label="Toggle navigation" name="search-input" data-search-index="../search.json" id="search-input" placeholder="Search for" autocomplete="off">
</form>

      <ul class="navbar-nav">
<li class="nav-item">
  <a class="external-link nav-link" href="https://github.com/joshwlambert/DAISIEprep/" aria-label="github">
    <span class="fab fa fab fa-github fa-lg"></span>
     
  </a>
</li>
      </ul>
</div>

    
  </div>
</nav><div class="container template-article">




<div class="row">
  <main id="main" class="col-md-9"><div class="page-header">
      <img src="../logo.png" class="logo" alt=""><h1>Extending DAISIEprep ASR models</h1>
            
      
      <small class="dont-index">Source: <a href="https://github.com/joshwlambert/DAISIEprep/blob/HEAD/vignettes/Extending_asr.Rmd" class="external-link"><code>vignettes/Extending_asr.Rmd</code></a></small>
      <div class="d-none name"><code>Extending_asr.Rmd</code></div>
    </div>

    
    
<div class="sourceCode" id="cb1"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">library</a></span><span class="op">(</span><span class="va"><a href="https://github.com/joshwlambert/DAISIEprep" class="external-link">DAISIEprep</a></span><span class="op">)</span></span></code></pre></div>
<div class="section level2">
<h2 id="introduction">Introduction<a class="anchor" aria-label="anchor" href="#introduction"></a>
</h2>
<p>In this tutorial we demonstrate how users can perform ancestral state
reconstruction using the functions implemented in
<code>DAISIEprep</code>, or, alternatively, by importing ancestral range
reconstructions obtained using methods from other packages.</p>
</div>
<div class="section level2">
<h2 id="using-daisiepreps-min-and-asr-algorithms">Using <code>DAISIEprep</code>’s <code>min</code> and
<code>asr</code> algorithms<a class="anchor" aria-label="anchor" href="#using-daisiepreps-min-and-asr-algorithms"></a>
</h2>
<p>The core feature of <code>DAISIEprep</code> is the function
<code><a href="../reference/extract_island_species.html">extract_island_species()</a></code>, which allows one to extract the
island data expected as input by <code>DAISIE</code>’s functions from a
phylogeny with data regarding presence / absence of each present-day
species from the island. The function automatically delineates and
extracts clades formed by island species within a complete mainland +
island phylogeny, and attempt to estimate the age of colonisation for
each of these clades.</p>
<p>The default option of the function is the <code>min</code> algorithm,
which performs the data extraction in a manner consistent with DAISIE’s
assumptions. However, there may be cases where it is not desirable to
use this algorithm, particularly if some of DAISIE’s assumptions are at
odds with the clade at hand. For example, consider the following
tree:</p>
<p>We have an island clade, comprising species a-f, except for species e
which is absent from the island. A parsimonious explanation for this
distribution would be a unique colonisation event before the split
between the f and a-e lineages, with the island population of species f
not diverging from its mainland ancestor, and species e jumping back to
the continent. This would result in a single island clade with a unique
colonisation time. Yet if we run <code>extract_island_species</code>
through this phylogeny:</p>
<div class="sourceCode" id="cb2"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="../reference/extract_island_species.html">extract_island_species</a></span><span class="op">(</span><span class="va">phylod</span>, extraction_method <span class="op">=</span> <span class="st">"min"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Class:  Island_tbl </span></span>
<span><span class="co">#&gt;   clade_name     status missing_species   col_time col_max_age branching_times</span></span>
<span><span class="co">#&gt; 1     bird_a    endemic               0 0.09053720       FALSE              NA</span></span>
<span><span class="co">#&gt; 2     bird_b    endemic               0 0.06146596       FALSE    0.016781....</span></span>
<span><span class="co">#&gt; 3     bird_d    endemic               0 0.05480904       FALSE              NA</span></span>
<span><span class="co">#&gt; 4     bird_f nonendemic               0 0.58496106       FALSE              NA</span></span>
<span><span class="co">#&gt;   min_age      species clade_type</span></span>
<span><span class="co">#&gt; 1      NA       bird_a          1</span></span>
<span><span class="co">#&gt; 2      NA bird_b, ....          1</span></span>
<span><span class="co">#&gt; 3      NA       bird_d          1</span></span>
<span><span class="co">#&gt; 4      NA       bird_f          1</span></span></code></pre></div>
<p>The algorithm estimates four independent colonisation events. This is
because the <code>min</code> algorithm assumes no back-colonisation
(from island to mainland), such that the presence of mainland-only
species e inside the island-only clade can only be accommodated by the
lineage staying on the mainland until the present, with at least three
colonisation events leading to species a-d. DAISIE would also not
consider a colonisation time before the (a-e)-f split, as any
cladogenetic event taking place on the island is assumed to lead to
strictly endemic lineages <span class="citation">(Valente et al.
2015)</span>, while lineage f maintains a population on the
mainland.</p>
<p>For such cases where the phylogeny is at odds with the process
considered by DAISIE, one may wish to resort to other trait evolution /
biogeography model to estimate when and how many times the island was
colonised. This requires performing ancestral state reconstruction, to
estimate the endemicity status of each internal node in the phylogeny.
<code>extract_island_species</code> offers the means to extract island
data from a phylogeny with completed node data, setting argument
<code>extraction_method = "asr"</code>.</p>
<p>The methods that <code>DAISIEprep</code> provides to run ancestral
state reconstruction (ASR) are parsimony and the Markov model (Mk) using
functionality from the R package <code>castor</code> <span class="citation">(Louca and Doebeli 2018)</span>. These are provided as
standard in the <code><a href="../reference/add_asr_node_states.html">DAISIEprep::add_asr_node_states()</a></code> function
to easily allow a user to run a quick reconstruction of the internal
nodes’ endemicity status. The parsimony and the Mk model provide simple
models that have been widely used in evolutionary biology since their
development. For details on the parsimony method see documentation for
<code><a href="https://rdrr.io/pkg/castor/man/asr_max_parsimony.html" class="external-link">castor::asr_max_parsimony()</a></code> and for details on the Mk model
see documentation for <code><a href="https://rdrr.io/pkg/castor/man/asr_mk_model.html" class="external-link">castor::asr_mk_model()</a></code>.</p>
<p>Here we show the same example as in the Tutorial vignette to show how
both methods are implemented:</p>
<div class="sourceCode" id="cb3"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span></span>
<span>  <span class="fl">1</span>,</span>
<span>  kind <span class="op">=</span> <span class="st">"Mersenne-Twister"</span>,</span>
<span>  normal.kind <span class="op">=</span> <span class="st">"Inversion"</span>,</span>
<span>  sample.kind <span class="op">=</span> <span class="st">"Rejection"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">phylo</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/rtree.html" class="external-link">rcoal</a></span><span class="op">(</span><span class="fl">10</span><span class="op">)</span></span>
<span></span>
<span><span class="va">phylo</span><span class="op">$</span><span class="va">tip.label</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bird_a"</span>, <span class="st">"bird_b"</span>, <span class="st">"bird_c"</span>, <span class="st">"bird_d"</span>, <span class="st">"bird_e"</span>, <span class="st">"bird_f"</span>,</span>
<span>                     <span class="st">"bird_g"</span>, <span class="st">"bird_h"</span>, <span class="st">"bird_i"</span>, <span class="st">"bird_j"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">phylo</span> <span class="op">&lt;-</span> <span class="fu">phylobase</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phylobase/man/phylo4-methods.html" class="external-link">phylo4</a></span><span class="op">(</span><span class="va">phylo</span><span class="op">)</span></span>
<span></span>
<span><span class="va">endemicity_status</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/sample.html" class="external-link">sample</a></span><span class="op">(</span></span>
<span>  x <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"not_present"</span>, <span class="st">"endemic"</span>, <span class="st">"nonendemic"</span><span class="op">)</span>, </span>
<span>  size <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="fu">phylobase</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phylobase/man/labels-methods.html" class="external-link">tipLabels</a></span><span class="op">(</span><span class="va">phylo</span><span class="op">)</span><span class="op">)</span>, </span>
<span>  replace <span class="op">=</span> <span class="cn">TRUE</span>,</span>
<span>  prob <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="fl">0.6</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">phylod</span> <span class="op">&lt;-</span> <span class="fu">phylobase</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phylobase/man/phylo4d-methods.html" class="external-link">phylo4d</a></span><span class="op">(</span><span class="va">phylo</span>, <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">endemicity_status</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># reconstruction using parsimony</span></span>
<span><span class="va">phylod_parsimony</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_asr_node_states.html">add_asr_node_states</a></span><span class="op">(</span></span>
<span>  phylod <span class="op">=</span> <span class="va">phylod</span>, </span>
<span>  asr_method <span class="op">=</span> <span class="st">"parsimony"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># reconstruction using Mk model</span></span>
<span><span class="va">phylod_parsimony</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/add_asr_node_states.html">add_asr_node_states</a></span><span class="op">(</span></span>
<span>  phylod <span class="op">=</span> <span class="va">phylod</span>, </span>
<span>  asr_method <span class="op">=</span> <span class="st">"Mk"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>For details on the internal workings of the
<code><a href="../reference/add_asr_node_states.html">add_asr_node_states()</a></code> function see appendix at the bottom
of this article.</p>
</div>
<div class="section level2">
<h2 id="using-ancestral-state-reconstruction-methods-from-other-packages">Using ancestral state reconstruction methods from other
packages<a class="anchor" aria-label="anchor" href="#using-ancestral-state-reconstruction-methods-from-other-packages"></a>
</h2>
<p>The <code>min</code> and <code>asr</code> algorithms are implemented
in <code>DAISIEprep</code>. However, there are many models developed for
the reconstruction of states (traits) on a phylogenetic tree available
in other R packages, and it may be more appropriate to use a different
type of model for the empirical group being studied. Just as R is
developed to allow for packages to <a href="https://cran.r-project.org/doc/manuals/r-release/R-exts.html" class="external-link">extend
the language</a>, <code>DAISIEprep</code> is designed to allow each
extension of ASR methods for incorporation with key functions
(e.g. <code><a href="../reference/extract_island_species.html">extract_island_species()</a></code>).</p>
<p>Here we give examples of three packages that can be used an
extensions: <code>diversitree</code>, <code>BioGeoBEARS</code> and
<code>corHMM</code>.</p>
<p><code>diversitree</code> <span class="citation">(FitzJohn
2012)</span> is a package containing a suite of State Speciation and
Extinction (SSE) model which can reconstruct ancestral states under a
model in which the rates of speciation, extinction and transition
between states all influence the reconstruction. These models prevent
the bias of having many species in a state because of high speciation
but a model, such as the Mk model, assumes it is due to high transition
rates into that state (see <span class="citation">Maddison and Knowles
(2006)</span>). The example we give uses the MuSSE model with a three
states (island endemic, island non-endemic and not present on the
island), and the GeoSSE model that considers presence or absence from
two geographic areas (island and mainland). Other SSE models in
<code>diversitree</code> can be applied in the same manner.</p>
<p><code>BioGeoBEARS</code> <span class="citation">(Matzke 2013)</span>
is a widely used package that includes the DEC and DEC+J models of
biogeographic reconstruction. Therefore, it may be that people familiar
with these models want to apply them for extracting island colonisations
for DAISIE.</p>
<p>Lastly, <code>corHMM</code> <span class="citation">(Beaulieu et al.
2013)</span> is a package that implements a hidden markov model of
evolution, similar to the Mk model, but can better account for rate
heterogeneity by introducing hidden states into the model. Each model
can be argued for or against; with the choice influenced by the
taxonomic group being studied.</p>
</div>
<div class="section level2">
<h2 id="decj">DEC+J<a class="anchor" aria-label="anchor" href="#decj"></a>
</h2>
<p>We consider the following randomly generated phylogeny and tip
data:</p>
<p>As a first example, we consider the popular DEC
(Dispersal-Extinction-Cladogenesis) model <span class="citation">(Ree
and Smith 2008)</span> with founder-event speciation (DEC+J, <span class="citation">Matzke (2013)</span>), implemented in R in the
<strong>biogeobears</strong> package <span class="citation">(Matzke
2013)</span>.</p>
<div class="sourceCode" id="cb4"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw"><a href="https://rdrr.io/r/base/library.html" class="external-link">require</a></span><span class="op">(</span><span class="va"><a href="http://phylo.wikidot.com/biogeobears" class="external-link">BioGeoBEARS</a></span><span class="op">)</span></span>
<span><span class="co">#&gt; Loading required package: BioGeoBEARS</span></span></code></pre></div>
<p>BioGeoBEARS revolves around an object, <code>BioGeoBEARS_run</code>,
which stores input data, the structure of the model to optimise, and
control parameters for optimisation.</p>
<div class="sourceCode" id="cb5"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Default structure of the BioGeoBEARS object</span></span>
<span><span class="va">bgb_run</span> <span class="op">&lt;-</span> <span class="fu">BioGeoBEARS</span><span class="fu">::</span><span class="fu">define_BioGeoBEARS_run</span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">bgb_run</span></span>
<span><span class="co">#&gt; $geogfn</span></span>
<span><span class="co">#&gt; [1] "/home/runner/work/_temp/Library/BioGeoBEARS/extdata/Psychotria_geog.data"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $trfn</span></span>
<span><span class="co">#&gt; [1] "/home/runner/work/_temp/Library/BioGeoBEARS/extdata/Psychotria_5.2.newick"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $abbr</span></span>
<span><span class="co">#&gt; [1] "default"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $description</span></span>
<span><span class="co">#&gt; [1] "defaults"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $BioGeoBEARS_model_object</span></span>
<span><span class="co">#&gt; An object of class "BioGeoBEARS_model"</span></span>
<span><span class="co">#&gt; Slot "params_table":</span></span>
<span><span class="co">#&gt;          type    init      min      max     est                note</span></span>
<span><span class="co">#&gt; d        free 0.01000  1.0e-12  5.00000 0.01000               works</span></span>
<span><span class="co">#&gt; e        free 0.01000  1.0e-12  5.00000 0.01000               works</span></span>
<span><span class="co">#&gt; a       fixed 0.00000  1.0e-12  5.00000 0.00000               works</span></span>
<span><span class="co">#&gt; b       fixed 1.00000  1.0e-12  1.00000 1.00000 non-stratified only</span></span>
<span><span class="co">#&gt; x       fixed 0.00000 -2.5e+00  2.50000 0.00000               works</span></span>
<span><span class="co">#&gt; n       fixed 0.00000 -1.0e+01 10.00000 0.00000               works</span></span>
<span><span class="co">#&gt; w       fixed 1.00000 -1.0e+01 10.00000 1.00000               works</span></span>
<span><span class="co">#&gt; u       fixed 0.00000 -1.0e+01 10.00000 0.00000               works</span></span>
<span><span class="co">#&gt; j       fixed 0.00000  1.0e-05  2.99999 0.00000               works</span></span>
<span><span class="co">#&gt; ysv       3-j 2.99999  1.0e-05  3.00000 2.99999               works</span></span>
<span><span class="co">#&gt; ys    ysv*2/3 1.99999  1.0e-05  2.00000 1.99999               works</span></span>
<span><span class="co">#&gt; y     ysv*1/3 1.00000  1.0e-05  1.00000 1.00000               works</span></span>
<span><span class="co">#&gt; s     ysv*1/3 1.00000  1.0e-05  1.00000 1.00000               works</span></span>
<span><span class="co">#&gt; v     ysv*1/3 1.00000  1.0e-05  1.00000 1.00000               works</span></span>
<span><span class="co">#&gt; mx01    fixed 0.00010  1.0e-04  0.99990 0.00010               works</span></span>
<span><span class="co">#&gt; mx01j    mx01 0.00010  1.0e-04  0.99990 0.00010               works</span></span>
<span><span class="co">#&gt; mx01y    mx01 0.00010  1.0e-04  0.99990 0.00010               works</span></span>
<span><span class="co">#&gt; mx01s    mx01 0.00010  1.0e-04  0.99990 0.00010               works</span></span>
<span><span class="co">#&gt; mx01v    mx01 0.00010  1.0e-04  0.99990 0.00010               works</span></span>
<span><span class="co">#&gt; mx01r   fixed 0.50000  1.0e-04  0.99990 0.50000                  no</span></span>
<span><span class="co">#&gt; mf      fixed 0.10000  5.0e-03  0.99500 0.10000                 yes</span></span>
<span><span class="co">#&gt; dp      fixed 1.00000  5.0e-03  0.99500 1.00000                 yes</span></span>
<span><span class="co">#&gt; fdp     fixed 0.00000  5.0e-03  0.99500 0.00000                 yes</span></span>
<span><span class="co">#&gt;                                                                        desc</span></span>
<span><span class="co">#&gt; d                         anagenesis: rate of 'dispersal' (range expansion)</span></span>
<span><span class="co">#&gt; e                      anagenesis: rate of 'extinction' (range contraction)</span></span>
<span><span class="co">#&gt; a           anagenesis: rate of range-switching (i.e. for a standard char.)</span></span>
<span><span class="co">#&gt; b                                    anagenesis: exponent on branch lengths</span></span>
<span><span class="co">#&gt; x                                   exponent on distance (modifies d, j, a)</span></span>
<span><span class="co">#&gt; n                     exponent on environmental distance (modifies d, j, a)</span></span>
<span><span class="co">#&gt; w               exponent on manual dispersal multipliers (modifies d, j, a)</span></span>
<span><span class="co">#&gt; u            anagenesis: exponent on extinction risk with area (modifies e)</span></span>
<span><span class="co">#&gt; j                 cladogenesis: relative per-event weight of jump dispersal</span></span>
<span><span class="co">#&gt; ysv                                                     cladogenesis: y+s+v</span></span>
<span><span class="co">#&gt; ys                                                        cladogenesis: y+s</span></span>
<span><span class="co">#&gt; y       cladogenesis: relative per-event weight of sympatry (range-copying)</span></span>
<span><span class="co">#&gt; s              cladogenesis: relative per-event weight of subset speciation</span></span>
<span><span class="co">#&gt; v           cladogenesis: relative per-event weight of vicariant speciation</span></span>
<span><span class="co">#&gt; mx01                  cladogenesis: controls range size of smaller daughter</span></span>
<span><span class="co">#&gt; mx01j                 cladogenesis: controls range size of smaller daughter</span></span>
<span><span class="co">#&gt; mx01y                 cladogenesis: controls range size of smaller daughter</span></span>
<span><span class="co">#&gt; mx01s                 cladogenesis: controls range size of smaller daughter</span></span>
<span><span class="co">#&gt; mx01v                 cladogenesis: controls range size of smaller daughter</span></span>
<span><span class="co">#&gt; mx01r                       root: controls range size probabilities of root</span></span>
<span><span class="co">#&gt; mf                         mean frequency of truly sampling OTU of interest</span></span>
<span><span class="co">#&gt; dp                 detection probability per true sample of OTU of interest</span></span>
<span><span class="co">#&gt; fdp   false detection of OTU probability per true taphonomic control sample</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $timesfn</span></span>
<span><span class="co">#&gt; [1] NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $distsfn</span></span>
<span><span class="co">#&gt; [1] NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $dispersal_multipliers_fn</span></span>
<span><span class="co">#&gt; [1] NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $area_of_areas_fn</span></span>
<span><span class="co">#&gt; [1] NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $areas_allowed_fn</span></span>
<span><span class="co">#&gt; [1] NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $areas_adjacency_fn</span></span>
<span><span class="co">#&gt; [1] NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $detects_fn</span></span>
<span><span class="co">#&gt; [1] NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $controls_fn</span></span>
<span><span class="co">#&gt; [1] NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $max_range_size</span></span>
<span><span class="co">#&gt; [1] NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $force_sparse</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $use_detection_model</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $print_optim</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $printlevel</span></span>
<span><span class="co">#&gt; [1] 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $on_NaN_error</span></span>
<span><span class="co">#&gt; [1] -1e+50</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $wd</span></span>
<span><span class="co">#&gt; [1] "/home/runner/work/DAISIEprep/DAISIEprep/vignettes"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $num_cores_to_use</span></span>
<span><span class="co">#&gt; [1] NA</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $cluster_already_open</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $use_optimx</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $rescale_params</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $return_condlikes_table</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $calc_TTL_loglike_from_condlikes_table</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $calc_ancprobs</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $speedup</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $include_null_range</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $useAmbiguities</span></span>
<span><span class="co">#&gt; [1] FALSE</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $min_branchlength</span></span>
<span><span class="co">#&gt; [1] 1e-06</span></span></code></pre></div>
<p>Many elements of this list are only relevant for advanced options of
the model and can be ignored if these features are not used. For
example, BioGeoBEARS allows explicit modelling of the connectivity
between areas and time-dependent availability of the areas. In this
example, we focus on a simple dispersal scenario between two areas
(mainland and island), so these elements can be ignored. We direct the
interested user to the <a href="http://phylo.wikidot.com/biogeobears#toc17" class="external-link">relevant tutorial</a>
on the BioGeoBEARS website.</p>
<p>BioGeoBEARS expects at least two inputs, the phylogeny and the
biogeographic data, a matrix of tip states. Both must be supplied as
paths to files which will be read when the model is run.</p>
<p>The tree can be supplied in Newick or Nexus format, as a text
file.</p>
<div class="sourceCode" id="cb6"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">path_to_phylo</span> <span class="op">&lt;-</span> <span class="st">"../inst/extending_asr/biogeobears_ex_phylo.txt"</span></span>
<span><span class="va">phylo</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="va">phylod</span>, <span class="st">"phylo"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in asMethod(object): losing data while coercing phylo4d to phylo</span></span>
<span><span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/write.tree.html" class="external-link">write.tree</a></span><span class="op">(</span><span class="va">phylo</span>, file <span class="op">=</span> <span class="va">path_to_phylo</span><span class="op">)</span></span>
<span></span>
<span><span class="va">bgb_run</span><span class="op">$</span><span class="va">trfn</span> <span class="op">&lt;-</span> <span class="va">path_to_phylo</span></span></code></pre></div>
<p>Tip data must be supplied as a text file specifying presence/absence
of every tip in each area, in the format used by the <a href="https://evolution.genetics.washington.edu/phylip.html" class="external-link">PHYLIP</a>
sofware suite. We report the full specifications from the <a href="http://phylo.wikidot.com/biogeobears#script" class="external-link">BioGeoBEARS</a>
tutorial</p>
<pre><code><span><span class="co">#######################################################</span></span>
<span><span class="co"># Geography file</span></span>
<span><span class="co"># Notes:</span></span>
<span><span class="co"># 1. This is a PHYLIP-formatted file. This means that in the </span></span>
<span><span class="co">#    first line, </span></span>
<span><span class="co">#    - the 1st number equals the number of rows (species)</span></span>
<span><span class="co">#    - the 2nd number equals the number of columns (number of areas)</span></span>
<span><span class="co">#    - after a tab, put the areas in parentheses, with spaces: (A B C D)</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># 1.5. Example first line:</span></span>
<span><span class="co">#    10    4    (A B C D)</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># 2. The second line, and subsequent lines:</span></span>
<span><span class="co">#    speciesA    0110</span></span>
<span><span class="co">#    speciesB    0111</span></span>
<span><span class="co">#    speciesC    0001</span></span>
<span><span class="co">#         ...</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># 2.5a. This means a TAB between the species name and the area 0/1s</span></span>
<span><span class="co"># 2.5b. This also means NO SPACE AND NO TAB between the area 0/1s.</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># 3. See example files at:</span></span>
<span><span class="co">#    http://phylo.wikidot.com/biogeobears#files</span></span>
<span><span class="co"># </span></span>
<span><span class="co"># 4. Make you understand what a PLAIN-TEXT EDITOR is:</span></span>
<span><span class="co">#    http://phylo.wikidot.com/biogeobears#texteditors</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># 3. The PHYLIP format is the same format used for C++ LAGRANGE geography files.</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># 4. All names in the geography file must match names in the phylogeny file.</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># 5. DON'T USE SPACES IN SPECIES NAMES, USE E.G. "_"</span></span>
<span><span class="co">#</span></span>
<span><span class="co"># 6. Operational taxonomic units (OTUs) should ideally be phylogenetic lineages, </span></span>
<span><span class="co">#    i.e. genetically isolated populations.  These may or may not be identical </span></span>
<span><span class="co">#    with species.  You would NOT want to just use specimens, as each specimen </span></span>
<span><span class="co">#    automatically can only live in 1 area, which will typically favor DEC+J </span></span>
<span><span class="co">#    models.  This is fine if the species/lineages really do live in single areas,</span></span>
<span><span class="co">#    but you wouldn't want to assume this without thinking about it at least. </span></span>
<span><span class="co">#    In summary, you should collapse multiple specimens into species/lineages if </span></span>
<span><span class="co">#    data indicates they are the same genetic population.</span></span>
<span><span class="co">######################################################</span></span></code></pre>
<p>For convenience, we have included a function that writes this file
from a <code>phylod</code> object for the simple mainland-island
case.</p>
<div class="sourceCode" id="cb8"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">path_to_biogeo</span> <span class="op">&lt;-</span> <span class="st">"../inst/extending_asr/biogeobears_ex_biogeo.txt"</span></span>
<span><span class="fu"><a href="../reference/write_phylip_biogeo_file.html">write_phylip_biogeo_file</a></span><span class="op">(</span><span class="va">phylod</span>, <span class="va">path_to_biogeo</span><span class="op">)</span></span>
<span><span class="fu">BioGeoBEARS</span><span class="fu">::</span><span class="fu">getranges_from_LagrangePHYLIP</span><span class="op">(</span><span class="va">path_to_biogeo</span><span class="op">)</span></span>
<span><span class="co">#&gt; An object of class "tipranges"</span></span>
<span><span class="co">#&gt; numeric(0)</span></span>
<span><span class="co">#&gt; Slot "df":</span></span>
<span><span class="co">#&gt;        M I</span></span>
<span><span class="co">#&gt; bird_a 0 1</span></span>
<span><span class="co">#&gt; bird_b 0 1</span></span>
<span><span class="co">#&gt; bird_c 0 1</span></span>
<span><span class="co">#&gt; bird_d 0 1</span></span>
<span><span class="co">#&gt; bird_e 0 1</span></span>
<span><span class="co">#&gt; bird_f 0 1</span></span>
<span><span class="co">#&gt; bird_g 1 1</span></span>
<span><span class="co">#&gt; bird_h 1 0</span></span>
<span><span class="co">#&gt; bird_i 0 1</span></span>
<span><span class="co">#&gt; bird_j 1 0</span></span>
<span><span class="co">#&gt; bird_k 1 0</span></span>
<span><span class="co">#&gt; bird_l 1 0</span></span>
<span><span class="co">#&gt; bird_m 0 1</span></span>
<span><span class="co">#&gt; bird_n 0 1</span></span>
<span><span class="co">#&gt; bird_o 0 1</span></span>
<span><span class="co">#&gt; bird_p 1 0</span></span>
<span><span class="co">#&gt; bird_q 1 0</span></span>
<span><span class="co">#&gt; bird_r 1 0</span></span>
<span><span class="co">#&gt; bird_s 1 0</span></span>
<span><span class="co">#&gt; bird_t 1 0</span></span>
<span><span class="va">bgb_run</span><span class="op">$</span><span class="va">geogfn</span> <span class="op">&lt;-</span> <span class="va">path_to_biogeo</span></span></code></pre></div>
<p>While we were at it, we have nested this function in
<code><a href="../reference/write_biogeobears_input.html">write_biogeobears_input()</a></code>, to prepare both this file and
the Newick file above in one command</p>
<div class="sourceCode" id="cb9"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">path_to_phylo</span> <span class="op">&lt;-</span> <span class="st">"../inst/extending_asr/biogeobears_ex_phylo.txt"</span></span>
<span><span class="va">path_to_biogeo</span> <span class="op">&lt;-</span> <span class="st">"../inst/extending_asr/biogeobears_ex_biogeo.txt"</span></span>
<span><span class="fu"><a href="../reference/write_biogeobears_input.html">write_biogeobears_input</a></span><span class="op">(</span><span class="va">phylod</span>, <span class="va">path_to_phylo</span>, <span class="va">path_to_biogeo</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in asMethod(object): losing data while coercing phylo4d to phylo</span></span>
<span></span>
<span><span class="va">bgb_run</span><span class="op">$</span><span class="va">trfn</span> <span class="op">&lt;-</span> <span class="va">path_to_phylo</span></span>
<span><span class="va">bgb_run</span><span class="op">$</span><span class="va">geogfn</span> <span class="op">&lt;-</span> <span class="va">path_to_biogeo</span></span></code></pre></div>
<p>The structure of the model is contained in
<code>BioGeoBEARS_model_object</code>. This is simply a table that
contains the status (fixed or free), values (initial, min/max boundaries
and estimated value if free) and documentation of each parameter of the
supermodel.</p>
<div class="sourceCode" id="cb10"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bgb_run</span><span class="op">$</span><span class="va">BioGeoBEARS_model_object</span></span>
<span><span class="co">#&gt; An object of class "BioGeoBEARS_model"</span></span>
<span><span class="co">#&gt; Slot "params_table":</span></span>
<span><span class="co">#&gt;          type    init      min      max     est                note</span></span>
<span><span class="co">#&gt; d        free 0.01000  1.0e-12  5.00000 0.01000               works</span></span>
<span><span class="co">#&gt; e        free 0.01000  1.0e-12  5.00000 0.01000               works</span></span>
<span><span class="co">#&gt; a       fixed 0.00000  1.0e-12  5.00000 0.00000               works</span></span>
<span><span class="co">#&gt; b       fixed 1.00000  1.0e-12  1.00000 1.00000 non-stratified only</span></span>
<span><span class="co">#&gt; x       fixed 0.00000 -2.5e+00  2.50000 0.00000               works</span></span>
<span><span class="co">#&gt; n       fixed 0.00000 -1.0e+01 10.00000 0.00000               works</span></span>
<span><span class="co">#&gt; w       fixed 1.00000 -1.0e+01 10.00000 1.00000               works</span></span>
<span><span class="co">#&gt; u       fixed 0.00000 -1.0e+01 10.00000 0.00000               works</span></span>
<span><span class="co">#&gt; j       fixed 0.00000  1.0e-05  2.99999 0.00000               works</span></span>
<span><span class="co">#&gt; ysv       3-j 2.99999  1.0e-05  3.00000 2.99999               works</span></span>
<span><span class="co">#&gt; ys    ysv*2/3 1.99999  1.0e-05  2.00000 1.99999               works</span></span>
<span><span class="co">#&gt; y     ysv*1/3 1.00000  1.0e-05  1.00000 1.00000               works</span></span>
<span><span class="co">#&gt; s     ysv*1/3 1.00000  1.0e-05  1.00000 1.00000               works</span></span>
<span><span class="co">#&gt; v     ysv*1/3 1.00000  1.0e-05  1.00000 1.00000               works</span></span>
<span><span class="co">#&gt; mx01    fixed 0.00010  1.0e-04  0.99990 0.00010               works</span></span>
<span><span class="co">#&gt; mx01j    mx01 0.00010  1.0e-04  0.99990 0.00010               works</span></span>
<span><span class="co">#&gt; mx01y    mx01 0.00010  1.0e-04  0.99990 0.00010               works</span></span>
<span><span class="co">#&gt; mx01s    mx01 0.00010  1.0e-04  0.99990 0.00010               works</span></span>
<span><span class="co">#&gt; mx01v    mx01 0.00010  1.0e-04  0.99990 0.00010               works</span></span>
<span><span class="co">#&gt; mx01r   fixed 0.50000  1.0e-04  0.99990 0.50000                  no</span></span>
<span><span class="co">#&gt; mf      fixed 0.10000  5.0e-03  0.99500 0.10000                 yes</span></span>
<span><span class="co">#&gt; dp      fixed 1.00000  5.0e-03  0.99500 1.00000                 yes</span></span>
<span><span class="co">#&gt; fdp     fixed 0.00000  5.0e-03  0.99500 0.00000                 yes</span></span>
<span><span class="co">#&gt;                                                                        desc</span></span>
<span><span class="co">#&gt; d                         anagenesis: rate of 'dispersal' (range expansion)</span></span>
<span><span class="co">#&gt; e                      anagenesis: rate of 'extinction' (range contraction)</span></span>
<span><span class="co">#&gt; a           anagenesis: rate of range-switching (i.e. for a standard char.)</span></span>
<span><span class="co">#&gt; b                                    anagenesis: exponent on branch lengths</span></span>
<span><span class="co">#&gt; x                                   exponent on distance (modifies d, j, a)</span></span>
<span><span class="co">#&gt; n                     exponent on environmental distance (modifies d, j, a)</span></span>
<span><span class="co">#&gt; w               exponent on manual dispersal multipliers (modifies d, j, a)</span></span>
<span><span class="co">#&gt; u            anagenesis: exponent on extinction risk with area (modifies e)</span></span>
<span><span class="co">#&gt; j                 cladogenesis: relative per-event weight of jump dispersal</span></span>
<span><span class="co">#&gt; ysv                                                     cladogenesis: y+s+v</span></span>
<span><span class="co">#&gt; ys                                                        cladogenesis: y+s</span></span>
<span><span class="co">#&gt; y       cladogenesis: relative per-event weight of sympatry (range-copying)</span></span>
<span><span class="co">#&gt; s              cladogenesis: relative per-event weight of subset speciation</span></span>
<span><span class="co">#&gt; v           cladogenesis: relative per-event weight of vicariant speciation</span></span>
<span><span class="co">#&gt; mx01                  cladogenesis: controls range size of smaller daughter</span></span>
<span><span class="co">#&gt; mx01j                 cladogenesis: controls range size of smaller daughter</span></span>
<span><span class="co">#&gt; mx01y                 cladogenesis: controls range size of smaller daughter</span></span>
<span><span class="co">#&gt; mx01s                 cladogenesis: controls range size of smaller daughter</span></span>
<span><span class="co">#&gt; mx01v                 cladogenesis: controls range size of smaller daughter</span></span>
<span><span class="co">#&gt; mx01r                       root: controls range size probabilities of root</span></span>
<span><span class="co">#&gt; mf                         mean frequency of truly sampling OTU of interest</span></span>
<span><span class="co">#&gt; dp                 detection probability per true sample of OTU of interest</span></span>
<span><span class="co">#&gt; fdp   false detection of OTU probability per true taphonomic control sample</span></span></code></pre></div>
<p>BioGeoBEARS is indeed built as a supermodel which parameters can be
turned on or off to reproduce biogeographic models like DEC, DIVA,
BayArea and/or expand them.</p>
<p>See Fig. 1 in <span class="citation">Matzke (2013)</span> for an
overview of the supermodel and parameters:</p>
<div class="sourceCode" id="cb11"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">knitr</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/knitr/man/include_graphics.html" class="external-link">include_graphics</a></span><span class="op">(</span><span class="st">"http://phylo.wdfiles.com/local--files/biogeobears/BioGeoBEARS_supermodel.png"</span><span class="op">)</span></span></code></pre></div>
<p><img src="http://phylo.wdfiles.com/local--files/biogeobears/BioGeoBEARS_supermodel.png"><!-- --></p>
<p>Note that by default, all parameters but <em>d</em> and <em>e</em>
are turned off (i.e., fixed and set to a value such that they cause no
effect). That is, by default, <code>BioGeoBEARS_model_object</code>
specifies the DEC model.</p>
<p>For this example, we simply modify the model to make <em>j</em> a
free parameter, and thus turn the model into DEC+J.</p>
<div class="sourceCode" id="cb12"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># DEC -&gt; DEC+J</span></span>
<span><span class="va">bgb_run</span><span class="op">$</span><span class="va">BioGeoBEARS_model_object</span><span class="op">@</span><span class="va">params_table</span><span class="op">$</span><span class="va">desc</span><span class="op">[</span><span class="fl">9</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="st">"free"</span></span>
<span><span class="va">bgb_run</span><span class="op">$</span><span class="va">BioGeoBEARS_model_object</span><span class="op">@</span><span class="va">params_table</span><span class="op">$</span><span class="va">init</span><span class="op">[</span><span class="fl">9</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">0.01</span> <span class="co"># same value as d, e</span></span></code></pre></div>
<p>Some further controls:</p>
<div class="sourceCode" id="cb13"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">bgb_run</span><span class="op">$</span><span class="va">num_cores_to_use</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="co"># no default value on this one</span></span>
<span><span class="va">bgb_run</span><span class="op">$</span><span class="va">print_optim</span> <span class="op">&lt;-</span> <span class="cn">FALSE</span> <span class="co"># for the sake of the vignette</span></span></code></pre></div>
<p>Once everything is set up, it’s a good idea to check that the input
complies to the format expected by BioGeoBEARS with the provided
function. Then, we’re ready to run the optimisation:</p>
<div class="sourceCode" id="cb14"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="fu">BioGeoBEARS</span><span class="fu">::</span><span class="fu">check_BioGeoBEARS_run</span><span class="op">(</span><span class="va">bgb_run</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] TRUE</span></span>
<span><span class="va">res</span> <span class="op">&lt;-</span> <span class="fu">BioGeoBEARS</span><span class="fu">::</span><span class="fu">bears_optim_run</span><span class="op">(</span><span class="va">bgb_run</span><span class="op">)</span></span>
<span><span class="co">#&gt; [1] "Note: tipranges_to_tip_condlikes_of_data_on_each_state() is converting a states_list with (0-based) numbers to the equivalent areanames"</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Your computer has 2 cores.</span></span>
<span><span class="co">#&gt; [1] "parscale:"</span></span>
<span><span class="co">#&gt; [1] 1 1</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; NOTE: Before running optimx(), here is a test calculation of the data likelihood</span></span>
<span><span class="co">#&gt; using calc_loglike_for_optim() on initial parameter values, with printlevel=2...</span></span>
<span><span class="co">#&gt; if this crashes, the error messages are more helpful</span></span>
<span><span class="co">#&gt; than those from inside optimx().</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; calc_loglike_for_optim() on initial parameters loglike=-26.4272</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Calculation of likelihood on initial parameters: successful.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Now starting Maximum Likelihood (ML) parameter optimization with optimx()...</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Printing any warnings() that occurred during calc_loglike_for_optim():</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; NULL</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Results of optimx_scalecheck() below. Note: sometimes rescaling parameters may be helpful for ML searches, when the parameters have much different absolute sizes. This can be attempted by setting BioGeoBEARS_run_object$rescale_params = TRUE.</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $lpratio</span></span>
<span><span class="co">#&gt; [1] 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; $lbratio</span></span>
<span><span class="co">#&gt; [1] 0</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; Maximizing -- use negfn and neggr</span></span>
<span><span class="co">#&gt; Warning in (function (npt = min(n + 2L, 2L * n), rhobeg = NA, rhoend = NA, :</span></span>
<span><span class="co">#&gt; unused control arguments ignored</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt; This is the output from optim, optimx, or GenSA. Check the help on those functions to</span></span>
<span><span class="co">#&gt; interpret this output and check for convergence issues:</span></span>
<span><span class="co">#&gt; </span></span>
<span><span class="co">#&gt;               p1           p2   value fevals gevals niter convcode  kkt1 kkt2</span></span>
<span><span class="co">#&gt; bobyqa 0.3984246 1.106577e-08 -19.384     65     NA    NA        0 FALSE TRUE</span></span>
<span><span class="co">#&gt;        xtime</span></span>
<span><span class="co">#&gt; bobyqa 0.846</span></span></code></pre></div>
<p>We do get a warning, but <a href="http://phylo.wikidot.com/biogeobears-warnings-to-ignore-mostly#unused_control_arguments" class="external-link">apparently</a>
this can be ignored.</p>
<p>Marginal probabilities for the ancestral states appear to be found in
element <code>$ML_marginal_prob_each_state_at_branch_top_AT_node</code>
of the output (from what I infer from the code of the <a href="https://github.com/nmatzke/BioGeoBEARS/blob/7b16e5263e91389d8e16b7bead4437d39b8be5bc/R/BioGeoBEARS_plots_v1.R#L432" class="external-link">plotting
function code</a>). Columns respectively correspond to the posterior
probability of the node being present in the first area only (here,
mainland, i.e. not present), second area only (island, endemic), or both
areas (widespread state, nonendemic). Rows correspond to nodes
(including tips!) and their order is the same as in the tree object
<em>read from the input Newick file</em>, which may differ from the
original tree. To identify the probability associated to each node we
need to match the rows with the corresponding tip labels.</p>
<div class="sourceCode" id="cb15"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Extract probabilities</span></span>
<span><span class="va">asr_likelihoods</span> <span class="op">&lt;-</span> <span class="va">res</span><span class="op">$</span><span class="va">ML_marginal_prob_each_state_at_branch_top_AT_node</span><span class="op">[</span>,<span class="op">-</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="co"># Need to find which tip label matches each row</span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/read.tree.html" class="external-link">read.tree</a></span><span class="op">(</span><span class="va">res</span><span class="op">$</span><span class="va">inputs</span><span class="op">$</span><span class="va">trfn</span><span class="op">)</span></span>
<span><span class="va">tip_labels</span> <span class="op">&lt;-</span> <span class="va">tree</span><span class="op">$</span><span class="va">tip.label</span></span>
<span><span class="va">node_labels</span> <span class="op">&lt;-</span> <span class="va">tree</span><span class="op">$</span><span class="va">node.label</span></span>
<span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/NULL.html" class="external-link">is.null</a></span><span class="op">(</span><span class="va">node_labels</span><span class="op">)</span><span class="op">)</span> <span class="va">node_labels</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="fu"><a href="https://rdrr.io/r/base/length.html" class="external-link">length</a></span><span class="op">(</span><span class="va">tip_labels</span><span class="op">)</span> <span class="op">-</span> <span class="fl">1</span><span class="op">)</span></span>
<span><span class="va">asr_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  labels <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="va">tip_labels</span>, <span class="va">node_labels</span><span class="op">)</span>,</span>
<span>  not_present_prob <span class="op">=</span> <span class="va">asr_likelihoods</span><span class="op">[</span>,<span class="fl">1</span><span class="op">]</span>,</span>
<span>  endemic_prob <span class="op">=</span> <span class="va">asr_likelihoods</span><span class="op">[</span>,<span class="fl">2</span><span class="op">]</span>,</span>
<span>  nondendemic_prob <span class="op">=</span> <span class="va">asr_likelihoods</span><span class="op">[</span>,<span class="fl">3</span><span class="op">]</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">head</a></span><span class="op">(</span><span class="va">asr_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;   labels not_present_prob endemic_prob nondendemic_prob</span></span>
<span><span class="co">#&gt; 1 bird_d                0            1                0</span></span>
<span><span class="co">#&gt; 2 bird_e                0            1                0</span></span>
<span><span class="co">#&gt; 3 bird_c                0            1                0</span></span>
<span><span class="co">#&gt; 4 bird_a                0            1                0</span></span>
<span><span class="co">#&gt; 5 bird_b                0            1                0</span></span>
<span><span class="co">#&gt; 6 bird_f                0            1                0</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/utils/head.html" class="external-link">tail</a></span><span class="op">(</span><span class="va">asr_df</span><span class="op">)</span></span>
<span><span class="co">#&gt;    labels not_present_prob endemic_prob nondendemic_prob</span></span>
<span><span class="co">#&gt; 34   &lt;NA&gt;     6.137318e-13 2.401258e-02     9.759874e-01</span></span>
<span><span class="co">#&gt; 35   &lt;NA&gt;     9.474826e-11 1.180088e-10     1.000000e+00</span></span>
<span><span class="co">#&gt; 36   &lt;NA&gt;     2.535833e-22 1.000000e+00     2.671022e-11</span></span>
<span><span class="co">#&gt; 37   &lt;NA&gt;     1.000000e+00 8.344796e-19     1.776066e-08</span></span>
<span><span class="co">#&gt; 38   &lt;NA&gt;     1.000000e+00 8.802463e-23     4.458255e-10</span></span>
<span><span class="co">#&gt; 39   &lt;NA&gt;     1.000000e+00 1.529832e-25     9.339950e-11</span></span></code></pre></div>
<p>The code above has been wrapped in a utility function,
<code><a href="../reference/extract_biogeobears_ancestral_states_probs.html">extract_biogeobears_ancestral_states_probs()</a></code>.</p>
<div class="sourceCode" id="cb16"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">asr_df</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/extract_biogeobears_ancestral_states_probs.html">extract_biogeobears_ancestral_states_probs</a></span><span class="op">(</span><span class="va">res</span><span class="op">)</span></span></code></pre></div>
<p>The last step before extracting the island community from the tree is
to rule which state each node is in from the probabilities. We provide a
convenient function to do this from the output of the previous
function:</p>
<div class="sourceCode" id="cb17"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">endemicity_status</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select_endemicity_status.html">select_endemicity_status</a></span><span class="op">(</span><span class="va">asr_df</span>, method <span class="op">=</span> <span class="st">"max"</span><span class="op">)</span></span></code></pre></div>
<p>By default, (<code>method = "max"</code>), we simply select the state
with the highest probability (preferring the last in the event of a
tie). <code>method</code> can also be set to <code>"random"</code> to
sample states randomly based on the probabilities, which can be of use
if one desires to explore the sensibility of downstream DAISIE analyses
to the ancestral state reconstruction.</p>
<p>Finally, we recreate the data format expected by
<code><a href="../reference/extract_island_species.html">extract_island_species()</a></code> (one state column
<code>endemicity_status</code> for tips, one state column
<code>island_status</code> for internal nodes) and add it back to the
tree data with <code>phylobase</code>’s <code>phylo4d</code> class.</p>
<div class="sourceCode" id="cb18"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Add endemicity data</span></span>
<span><span class="va">nb_tips</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/summary.phylo.html" class="external-link">Ntip</a></span><span class="op">(</span><span class="va">tree</span><span class="op">)</span></span>
<span><span class="va">asr_df</span><span class="op">$</span><span class="va">label</span> <span class="op">&lt;-</span> <span class="cn">NULL</span> <span class="co"># drop label</span></span>
<span><span class="va">asr_df</span><span class="op">$</span><span class="va">endemicity_status</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">asr_df</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">asr_df</span><span class="op">$</span><span class="va">endemicity_status</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">nb_tips</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">endemicity_status</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">nb_tips</span><span class="op">]</span> </span>
<span><span class="va">asr_df</span><span class="op">$</span><span class="va">island_status</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/rep.html" class="external-link">rep</a></span><span class="op">(</span><span class="cn">NA</span>, <span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">asr_df</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">asr_df</span><span class="op">$</span><span class="va">island_status</span><span class="op">[</span><span class="op">(</span><span class="va">nb_tips</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">asr_df</span><span class="op">)</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="va">endemicity_status</span><span class="op">[</span><span class="op">(</span><span class="va">nb_tips</span> <span class="op">+</span> <span class="fl">1</span><span class="op">)</span><span class="op">:</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">nrow</a></span><span class="op">(</span><span class="va">asr_df</span><span class="op">)</span><span class="op">]</span> </span>
<span></span>
<span><span class="co"># Rebuild phylod with ancestral states</span></span>
<span><span class="va">phylod</span> <span class="op">&lt;-</span> <span class="fu">phylobase</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phylobase/man/phylo4d-methods.html" class="external-link">phylo4d</a></span><span class="op">(</span><span class="va">tree</span>, all.data <span class="op">=</span> <span class="va">asr_df</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_phylod.html">plot_phylod</a></span><span class="op">(</span><span class="va">phylod</span><span class="op">)</span></span></code></pre></div>
<p><img src="Extending_asr_files/figure-html/order_biogeobears_asr-1.png" width="700"></p>
<p>Voilà!</p>
<div class="sourceCode" id="cb19"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">island_clades</span> <span class="op">&lt;-</span> <span class="fu">DAISIEprep</span><span class="fu">::</span><span class="fu"><a href="../reference/extract_island_species.html">extract_island_species</a></span><span class="op">(</span></span>
<span>  phylod <span class="op">=</span> <span class="va">phylod</span>,</span>
<span>  extraction_method <span class="op">=</span> <span class="st">"asr"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">island_clades</span><span class="op">@</span><span class="va">island_tbl</span><span class="op">$</span><span class="va">species</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt;  [1] "bird_d" "bird_e" "bird_c" "bird_a" "bird_b" "bird_f" "bird_i" "bird_g"</span></span>
<span><span class="co">#&gt;  [9] "bird_o" "bird_m" "bird_n"</span></span>
<span><span class="va">island_clades</span><span class="op">@</span><span class="va">island_tbl</span><span class="op">$</span><span class="va">branching_times</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt;  [1] 2.12780098 0.38960848 0.30994875 0.16207082 0.13800731 0.10346994</span></span>
<span><span class="co">#&gt;  [7] 0.08406573 0.04477260 0.02586891 0.01218459</span></span></code></pre></div>
</div>
<div class="section level2">
<h2 id="musse">MuSSE<a class="anchor" aria-label="anchor" href="#musse"></a>
</h2>
<p>SSE-class models explicitly model the inter-dependency between the
evolution of a set of traits and evolutionary rates (speciation and
extinction). When traits are set to represent geographic areas, such
models can be used to model range evolution. For a mainland-island
system, we could model the endemicity status as a ternary trait:
endemic, non-endemic, or not present on the island.</p>
<p>Discrete-trait models with more than two states are modelled with
MuSSE <span class="citation">(FitzJohn 2012)</span>, which is
implemented in package <code>diversitree</code>.</p>
<p>To make the example straightforward and avoid non-convergence issues,
we use a tree simulated directly with MuSSE as an example, with three
states (mainland/island/both, or not present/endemic/nonendemic) and
arbitrary parameter values:</p>
<div class="sourceCode" id="cb20"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Simulate a tree under a MuSSE model,</span></span>
<span><span class="co"># with arbitrary initial parameter values</span></span>
<span><span class="va">pars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="co"># lambda1 lambda2 lambda3</span></span>
<span>  <span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>,</span>
<span>  <span class="co"># mu1 mu2 mu3</span></span>
<span>  <span class="fl">0.02</span>, <span class="fl">0.02</span>, <span class="fl">0.02</span>,</span>
<span>  <span class="co"># q12 q13 q21 q23 q31 q32</span></span>
<span>  <span class="fl">0</span>,  <span class="fl">0.1</span>,  <span class="fl">0</span>, <span class="fl">0</span>,  <span class="fl">0.1</span>, <span class="fl">0.1</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/Random.html" class="external-link">set.seed</a></span><span class="op">(</span><span class="fl">4</span><span class="op">)</span></span>
<span><span class="va">phylo</span> <span class="op">&lt;-</span> <span class="fu">diversitree</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/diversitree/man/simulate.html" class="external-link">trees</a></span><span class="op">(</span></span>
<span>  <span class="va">pars</span>,</span>
<span>  type <span class="op">=</span> <span class="st">"musse"</span>,</span>
<span>  max.taxa <span class="op">=</span> <span class="fl">20</span>,</span>
<span>  max.t <span class="op">=</span> <span class="cn">Inf</span>,</span>
<span>  x0 <span class="op">=</span> <span class="fl">1</span></span>
<span><span class="op">)</span><span class="op">[[</span><span class="fl">1</span><span class="op">]</span><span class="op">]</span></span>
<span></span>
<span><span class="co"># DAISIEprep requires binomial names</span></span>
<span><span class="va">ntips</span> <span class="op">&lt;-</span> <span class="fu">ape</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/ape/man/summary.phylo.html" class="external-link">Ntip</a></span><span class="op">(</span><span class="va">phylo</span><span class="op">)</span></span>
<span><span class="va">bird_names</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="st">"bird_"</span>, <span class="va">letters</span><span class="op">[</span><span class="fl">1</span><span class="op">:</span><span class="va">ntips</span><span class="op">]</span><span class="op">)</span></span>
<span><span class="va">phylo</span><span class="op">$</span><span class="va">tip.label</span> <span class="op">&lt;-</span> <span class="va">bird_names</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/names.html" class="external-link">names</a></span><span class="op">(</span><span class="va">phylo</span><span class="op">$</span><span class="va">tip.state</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="va">bird_names</span></span>
<span></span>
<span><span class="co"># DAISIEprep requires a mainland outgroup</span></span>
<span><span class="va">outgroup_clade</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"bird_h"</span>, <span class="st">"bird_i"</span>, <span class="st">"bird_l"</span>, <span class="st">"bird_k"</span>, <span class="st">"bird_r"</span>, <span class="st">"bird_q"</span><span class="op">)</span></span>
<span><span class="va">phylo</span><span class="op">$</span><span class="va">tip.state</span><span class="op">[</span><span class="va">phylo</span><span class="op">$</span><span class="va">tip.label</span> <span class="op"><a href="https://rdrr.io/r/base/match.html" class="external-link">%in%</a></span> <span class="va">outgroup_clade</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span> <span class="co"># not_present</span></span></code></pre></div>
<p>The output phylogenetic trees come with simulated states attached at
the tips:</p>
<div class="sourceCode" id="cb21"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phylo</span><span class="op">$</span><span class="va">tip.state</span></span>
<span><span class="co">#&gt; bird_a bird_b bird_c bird_d bird_e bird_f bird_g bird_h bird_i bird_j bird_k </span></span>
<span><span class="co">#&gt;      1      3      1      3      2      2      2      1      1      1      1 </span></span>
<span><span class="co">#&gt; bird_l bird_m bird_n bird_o bird_p bird_q bird_r bird_s bird_t </span></span>
<span><span class="co">#&gt;      1      2      2      2      3      1      1      1      1</span></span></code></pre></div>
<p>Because we’re going to switch back and forth between those
single-digit states and the three endemicity status used by
<code>DAISIEprep</code>, we have included functions to switch easily
between the two:</p>
<div class="sourceCode" id="cb22"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">endemicity_status</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/sse_states_to_endemicity.html">sse_states_to_endemicity</a></span><span class="op">(</span><span class="va">phylo</span><span class="op">$</span><span class="va">tip.state</span>, sse_model <span class="op">=</span> <span class="st">"musse"</span><span class="op">)</span></span>
<span><span class="va">phylod</span> <span class="op">&lt;-</span> <span class="fu">phylobase</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phylobase/man/phylo4d-methods.html" class="external-link">phylo4d</a></span><span class="op">(</span><span class="va">phylo</span>, <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">endemicity_status</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_phylod.html">plot_phylod</a></span><span class="op">(</span><span class="va">phylod</span><span class="op">)</span></span></code></pre></div>
<p><img src="Extending_asr_files/figure-html/musse_states_to_endemic-1.png" width="700"></p>
<p>ASR with <code>diversitree</code> is done in three steps. The model
structure is first specified to create a likelihood function; then
parameters are optimised for this function; and finally the
probabilities of states of the internal nodes are determined with the
resulting maximum-likelihood model.</p>
<div class="sourceCode" id="cb23"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">states</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_sse_tip_states.html">get_sse_tip_states</a></span><span class="op">(</span><span class="va">phylod</span>, sse_model <span class="op">=</span> <span class="st">"musse"</span><span class="op">)</span></span>
<span><span class="va">states</span> <span class="co"># note that states must be named after tips</span></span>
<span><span class="co">#&gt; bird_a bird_b bird_c bird_d bird_e bird_f bird_g bird_h bird_i bird_j bird_k </span></span>
<span><span class="co">#&gt;      1      3      1      3      2      2      2      1      1      1      1 </span></span>
<span><span class="co">#&gt; bird_l bird_m bird_n bird_o bird_p bird_q bird_r bird_s bird_t </span></span>
<span><span class="co">#&gt;      1      2      2      2      3      1      1      1      1</span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="va">phylod</span>, <span class="st">"phylo"</span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in asMethod(object): losing data while coercing phylo4d to phylo</span></span>
<span></span>
<span><span class="co"># Create likelihood function</span></span>
<span><span class="va">lik_musse</span> <span class="op">&lt;-</span> <span class="fu">diversitree</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/diversitree/man/make.musse.html" class="external-link">make.musse</a></span><span class="op">(</span></span>
<span>  tree <span class="op">=</span> <span class="va">tree</span>, </span>
<span>  states <span class="op">=</span> <span class="va">states</span>, </span>
<span>  k <span class="op">=</span> <span class="fl">3</span></span>
<span><span class="op">)</span></span>
<span><span class="va">lik_musse</span></span>
<span><span class="co">#&gt; MuSSE likelihood function:</span></span>
<span><span class="co">#&gt;   * Parameter vector takes 12 elements:</span></span>
<span><span class="co">#&gt;      - lambda1, lambda2, lambda3, mu1, mu2, mu3, q12, q13, q21, q23,</span></span>
<span><span class="co">#&gt;        q31, q32</span></span>
<span><span class="co">#&gt;   * Function takes arguments (with defaults)</span></span>
<span><span class="co">#&gt;      - pars: Parameter vector</span></span>
<span><span class="co">#&gt;      - condition.surv [TRUE]: Condition likelihood on survial?</span></span>
<span><span class="co">#&gt;      - root [ROOT.OBS]: Type of root treatment</span></span>
<span><span class="co">#&gt;      - root.p [NULL]: Vector of root state probabilities</span></span>
<span><span class="co">#&gt;      - intermediates [FALSE]: Also return intermediate values?</span></span>
<span><span class="co">#&gt;   * Phylogeny with 20 tips and 19 nodes</span></span>
<span><span class="co">#&gt;      - Taxa: bird_a, bird_b, bird_c, bird_d, bird_e, bird_f, ...</span></span>
<span><span class="co">#&gt;   * Reference:</span></span>
<span><span class="co">#&gt;      - FitzJohn (submitted)</span></span>
<span><span class="co">#&gt; R definition:</span></span>
<span><span class="co">#&gt; function (pars, condition.surv = TRUE, root = ROOT.OBS, root.p = NULL, </span></span>
<span><span class="co">#&gt;     intermediates = FALSE)</span></span></code></pre></div>
<div class="sourceCode" id="cb24"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">fit_musse</span> <span class="op">&lt;-</span> <span class="fu">diversitree</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/diversitree/man/find.mle.html" class="external-link">find.mle</a></span><span class="op">(</span>func <span class="op">=</span> <span class="va">lik_musse</span>, x.init <span class="op">=</span> <span class="va">pars</span>, method <span class="op">=</span> <span class="st">"subplex"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># MuSSE ancestral state reconstructions under the ML model</span></span>
<span><span class="va">asr_musse</span> <span class="op">&lt;-</span> <span class="fu">diversitree</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/diversitree/man/asr.html" class="external-link">asr.marginal</a></span><span class="op">(</span>lik <span class="op">=</span> <span class="va">lik_musse</span>, pars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/coef.html" class="external-link">coef</a></span><span class="op">(</span><span class="va">fit_musse</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="va">asr_musse</span></span>
<span><span class="co">#&gt;              [,1]         [,2]         [,3]         [,4]         [,5]</span></span>
<span><span class="co">#&gt; [1,] 5.297008e-07 2.358182e-12 1.040125e-08 1.180891e-12 8.128492e-09</span></span>
<span><span class="co">#&gt; [2,] 9.999995e-01 1.000000e+00 1.000000e+00 1.000000e+00 1.000000e+00</span></span>
<span><span class="co">#&gt; [3,] 2.338548e-15 1.922647e-22 1.182962e-15 1.395899e-21 7.176299e-16</span></span>
<span><span class="co">#&gt;              [,6]         [,7]         [,8]         [,9]        [,10]</span></span>
<span><span class="co">#&gt; [1,] 2.449543e-13 5.181369e-14 9.975378e-01 9.999999e-01 1.339102e-13</span></span>
<span><span class="co">#&gt; [2,] 1.000000e+00 1.000000e+00 2.462179e-03 1.228694e-07 1.000000e+00</span></span>
<span><span class="co">#&gt; [3,] 3.144040e-15 3.691200e-15 1.784536e-09 7.700293e-13 1.101658e-21</span></span>
<span><span class="co">#&gt;             [,11]        [,12]        [,13]        [,14]        [,15]</span></span>
<span><span class="co">#&gt; [1,] 3.509243e-12 1.676497e-06 3.367431e-13 1.372973e-15 9.038228e-01</span></span>
<span><span class="co">#&gt; [2,] 1.000000e+00 9.999983e-01 1.000000e+00 1.000000e+00 9.617721e-02</span></span>
<span><span class="co">#&gt; [3,] 5.456285e-20 7.119444e-09 1.057905e-21 1.523247e-23 1.956740e-09</span></span>
<span><span class="co">#&gt;             [,16]        [,17]        [,18]        [,19]</span></span>
<span><span class="co">#&gt; [1,] 9.146031e-01 9.999140e-01 9.992933e-01 9.999998e-01</span></span>
<span><span class="co">#&gt; [2,] 8.539693e-02 8.604932e-05 7.067024e-04 2.129264e-07</span></span>
<span><span class="co">#&gt; [3,] 7.706430e-10 9.517947e-11 2.601569e-10 5.050529e-12</span></span></code></pre></div>
<p>We obtain a (pivoted) table of probabilites linked to each state,
from which we can select states as for the BioGeoBEARS example:</p>
<div class="sourceCode" id="cb25"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">asr_musse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">asr_musse</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">asr_musse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/paste.html" class="external-link">paste0</a></span><span class="op">(</span><span class="fu"><a href="../reference/all_endemicity_status.html">all_endemicity_status</a></span><span class="op">(</span><span class="op">)</span>, <span class="st">"_prob"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">island_status</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select_endemicity_status.html">select_endemicity_status</a></span><span class="op">(</span><span class="va">asr_musse</span>, method <span class="op">=</span> <span class="st">"max"</span><span class="op">)</span></span>
<span><span class="va">island_status</span></span>
<span><span class="co">#&gt;  [1] "endemic"     "endemic"     "endemic"     "endemic"     "endemic"    </span></span>
<span><span class="co">#&gt;  [6] "endemic"     "endemic"     "not_present" "not_present" "endemic"    </span></span>
<span><span class="co">#&gt; [11] "endemic"     "endemic"     "endemic"     "endemic"     "not_present"</span></span>
<span><span class="co">#&gt; [16] "not_present" "not_present" "not_present" "not_present"</span></span>
<span><span class="va">asr_musse</span><span class="op">$</span><span class="va">island_status</span> <span class="op">&lt;-</span> <span class="va">island_status</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">asr_musse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">phylobase</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phylobase/man/nodeId-methods.html" class="external-link">nodeId</a></span><span class="op">(</span><span class="va">phylod</span>, <span class="st">"internal"</span><span class="op">)</span></span>
<span></span>
<span><span class="va">phylod</span> <span class="op">&lt;-</span> <span class="fu">phylobase</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phylobase/man/addData-methods.html" class="external-link">addData</a></span><span class="op">(</span></span>
<span>  <span class="va">phylod</span>,</span>
<span>  node.data <span class="op">=</span> <span class="va">asr_musse</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_phylod.html">plot_phylod</a></span><span class="op">(</span><span class="va">phylod</span><span class="op">)</span></span></code></pre></div>
<p><img src="Extending_asr_files/figure-html/add_musse_asr_to_phylod-1.png" width="700"></p>
<p>From this completed tree we can extract the island clade:</p>
<div class="sourceCode" id="cb26"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">island_clades</span> <span class="op">&lt;-</span> <span class="fu">DAISIEprep</span><span class="fu">::</span><span class="fu"><a href="../reference/extract_island_species.html">extract_island_species</a></span><span class="op">(</span></span>
<span>  phylod <span class="op">=</span> <span class="va">phylod</span>,</span>
<span>  extraction_method <span class="op">=</span> <span class="st">"asr"</span></span>
<span><span class="op">)</span></span>
<span><span class="va">island_clades</span><span class="op">@</span><span class="va">island_tbl</span><span class="op">$</span><span class="va">species</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1] "bird_d" "bird_o" "bird_p" "bird_e" "bird_f" "bird_b" "bird_g" "bird_m"</span></span>
<span><span class="co">#&gt; [9] "bird_n"</span></span>
<span><span class="va">island_clades</span><span class="op">@</span><span class="va">island_tbl</span><span class="op">$</span><span class="va">branching_times</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1] 8.0047889 6.4518206 4.2518993 2.5503507 1.6635537 1.6299262 0.4224813</span></span>
<span><span class="co">#&gt; [8] 0.4212151</span></span></code></pre></div>
<p>SSE models in <code>diversitree</code> can be tuned further with
constraints that simplify the model. We can use this to make the model
closer to the assumptions of DAISIE. For example, it is not possible for
a species to jump directly to the island and disappear from the mainland
simultaneously; so that direct transitions from state 1 to 2 and back
should be forbidden. One could also assume mainland speciation to be
unlikely, resulting in zero speciation on the mainland and the same
speciation rate for endemic and nonendemic species:</p>
<div class="sourceCode" id="cb27"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">lik_musse_c</span> <span class="op">&lt;-</span> <span class="fu">diversitree</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/diversitree/man/constrain.html" class="external-link">constrain</a></span><span class="op">(</span></span>
<span>  <span class="va">lik_musse</span>, </span>
<span>  <span class="va">q12</span> <span class="op">~</span> <span class="fl">0</span>, <span class="va">q21</span> <span class="op">~</span> <span class="fl">0</span>, <span class="co"># no island-mainland jumps</span></span>
<span>  <span class="va">lambda1</span> <span class="op">~</span> <span class="fl">0</span>, <span class="va">lambda3</span> <span class="op">~</span> <span class="va">lambda2</span> <span class="co"># no mainland speciation</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Parameters must be removed accordingly</span></span>
<span><span class="va">pars_c</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="co"># lambda2</span></span>
<span>  <span class="fl">0.2</span>,</span>
<span>  <span class="co"># mu1 mu2 mu3</span></span>
<span>  <span class="fl">0.03</span>, <span class="fl">0.03</span>, <span class="fl">0.03</span>,</span>
<span>  <span class="co"># q13 q23 q31 q32</span></span>
<span>  <span class="fl">0</span>, <span class="fl">0.02</span>, <span class="fl">0</span>, <span class="fl">0.01</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="va">fit_musse_c</span> <span class="op">&lt;-</span> <span class="fu">diversitree</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/diversitree/man/find.mle.html" class="external-link">find.mle</a></span><span class="op">(</span>func <span class="op">=</span> <span class="va">lik_musse_c</span>, x.init <span class="op">=</span> <span class="va">pars_c</span>, method <span class="op">=</span> <span class="st">"subplex"</span><span class="op">)</span></span>
<span><span class="co"># Won't work because of poor parameter choice vs tip states</span></span></code></pre></div>
<p>The example above won’t succeed because of a mismatch between the
initial parameters value and the distribution of states at the tip, but
we include it for illustrating how one would implement constrains on a
<code>diversitree</code> model.</p>
<p>Instead of further constraining a multi-state model to emulate
dispersal and evolution between mainland and island, one may find more
appropriate to use an SSE model specifically dedicated to geographic
traits, i.e., GeoSSE.</p>
</div>
<div class="section level2">
<h2 id="geosse">GeoSSE<a class="anchor" aria-label="anchor" href="#geosse"></a>
</h2>
<p>GeoSSE <span class="citation">(Goldberg et al. 2011)</span> is a
special case of SSE models dedicated to the case where traits correspond
to presence or absence from two geographic areas, which is the case for
our mainland/island clade. Its most important feature is that instead of
considering the widespread state (mainland + island, that is nonendemic)
as a third separate category from the two single-area states (as in the
MuSSE example above), GeoSSE explicitly models it as corresponding to
presence in both areas, just as in BioGeoBEARS. Accordingly, dispersal
is unidirectional, from a single area towards the widespread state.
Extinction from an area the widespread state returns the species to the
remaining single area, while the same extinction while already in a
single area results in the true extinction of the species.</p>
<p>In fact, GeoSSE is the SSE implementation of the DEC model, the only
difference being that GeoSSE assumes an effect of areas themselves on
the rates of speciation and extinction; and thus on the branching
patterns observed in the tree [goldberg_phylogenetic_2011], while DEC
only maps geography along the branches of the tree.</p>
<p>We use the same tree as the previous example:</p>
<div class="sourceCode" id="cb28"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Rm all data but tip states</span></span>
<span><span class="va">phylod</span><span class="op">@</span><span class="va">data</span> <span class="op">&lt;-</span> <span class="va">phylod</span><span class="op">@</span><span class="va">data</span><span class="op">[</span><span class="fl">1</span><span class="op">]</span></span>
<span><span class="fu"><a href="../reference/plot_phylod.html">plot_phylod</a></span><span class="op">(</span><span class="va">phylod</span><span class="op">)</span></span></code></pre></div>
<p><img src="Extending_asr_files/figure-html/make_geosse_tree-1.png" width="700"></p>
<p>GeoSSE in <code>diversitree</code> has a different state encoding
than MuSSE: states are expected to be 0, 1, 2, with 0 corresponding to
the widespread state (that is, present on both mainland and island,
nonendemic) and the two other states corresponding to single geographic
areas. Hereafter, we follow 1 = mainland/not_present and 2 =
island/endemic.</p>
<p>The syntax is otherwise the same as for MuSSE (and other SSE models
implemented in <code>diversitree</code>).</p>
<div class="sourceCode" id="cb29"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">states</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/get_sse_tip_states.html">get_sse_tip_states</a></span><span class="op">(</span><span class="va">phylod</span>, sse_model <span class="op">=</span> <span class="st">"geosse"</span><span class="op">)</span></span>
<span><span class="va">states</span> <span class="co"># note the encoding differs from the MuSSE example</span></span>
<span><span class="co">#&gt; bird_a bird_b bird_c bird_d bird_e bird_f bird_g bird_h bird_i bird_j bird_k </span></span>
<span><span class="co">#&gt;      1      0      1      0      2      2      2      1      1      1      1 </span></span>
<span><span class="co">#&gt; bird_l bird_m bird_n bird_o bird_p bird_q bird_r bird_s bird_t </span></span>
<span><span class="co">#&gt;      1      2      2      2      0      1      1      1      1</span></span>
<span></span>
<span><span class="va">tree</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="va">phylod</span>, <span class="st">"phylo"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Create likelihood function</span></span>
<span><span class="va">lik_geosse</span> <span class="op">&lt;-</span> <span class="fu">diversitree</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/diversitree/man/make.geosse.html" class="external-link">make.geosse</a></span><span class="op">(</span></span>
<span>  tree <span class="op">=</span> <span class="va">tree</span>, </span>
<span>  states <span class="op">=</span> <span class="va">states</span></span>
<span><span class="op">)</span></span>
<span><span class="va">lik_geosse</span></span>
<span><span class="co">#&gt; GeoSSE likelihood function:</span></span>
<span><span class="co">#&gt;   * Parameter vector takes 7 elements:</span></span>
<span><span class="co">#&gt;      - sA, sB, sAB, xA, xB, dA, dB</span></span>
<span><span class="co">#&gt;   * Function takes arguments (with defaults)</span></span>
<span><span class="co">#&gt;      - pars: Parameter vector</span></span>
<span><span class="co">#&gt;      - condition.surv [TRUE]: Condition likelihood on survial?</span></span>
<span><span class="co">#&gt;      - root [ROOT.OBS]: Type of root treatment</span></span>
<span><span class="co">#&gt;      - root.p [NULL]: Vector of root state probabilities</span></span>
<span><span class="co">#&gt;      - intermediates [FALSE]: Also return intermediate values?</span></span>
<span><span class="co">#&gt;   * Phylogeny with 20 tips and 19 nodes</span></span>
<span><span class="co">#&gt;      - Taxa: bird_a, bird_b, bird_c, bird_d, bird_e, bird_f, ...</span></span>
<span><span class="co">#&gt;   * Reference:</span></span>
<span><span class="co">#&gt;      - Goldberg et al. (2011) doi:10.1093/sysbio/syr046</span></span>
<span><span class="co">#&gt; R definition:</span></span>
<span><span class="co">#&gt; function (pars, condition.surv = TRUE, root = ROOT.OBS, root.p = NULL, </span></span>
<span><span class="co">#&gt;     intermediates = FALSE)</span></span></code></pre></div>
<div class="sourceCode" id="cb30"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="co"># Initial GeoSSE parameter values</span></span>
<span><span class="va">pars</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span></span>
<span>  <span class="co"># Speciation: sA sB sAB</span></span>
<span>  <span class="fl">0.2</span>, <span class="fl">0.2</span>, <span class="fl">0.2</span>,</span>
<span>  <span class="co"># Extinction: xA xB</span></span>
<span>  <span class="fl">0.02</span>, <span class="fl">0.02</span>,</span>
<span>  <span class="co"># Dispersal: dA dB</span></span>
<span>  <span class="fl">0.1</span>, <span class="fl">0</span></span>
<span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Optimisation of the model's likelihood</span></span>
<span><span class="va">fit_geosse</span> <span class="op">&lt;-</span> <span class="fu">diversitree</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/diversitree/man/find.mle.html" class="external-link">find.mle</a></span><span class="op">(</span>func <span class="op">=</span> <span class="va">lik_geosse</span>, x.init <span class="op">=</span> <span class="va">pars</span>, method <span class="op">=</span> <span class="st">"subplex"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Ancestral state reconstructions under the ML model</span></span>
<span><span class="va">asr_geosse</span> <span class="op">&lt;-</span> <span class="fu">diversitree</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/diversitree/man/asr.html" class="external-link">asr.marginal</a></span><span class="op">(</span>lik <span class="op">=</span> <span class="va">lik_geosse</span>, pars <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/stats/coef.html" class="external-link">coef</a></span><span class="op">(</span><span class="va">fit_geosse</span><span class="op">)</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Select node states from marginal probabilities</span></span>
<span><span class="va">asr_geosse</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/t.html" class="external-link">t</a></span><span class="op">(</span><span class="va">asr_geosse</span><span class="op">)</span><span class="op">)</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">asr_geosse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"nonendemic_prob"</span>, <span class="st">"not_present_prob"</span>, <span class="st">"endemic_prob"</span><span class="op">)</span> <span class="co"># make sure to get the order right!</span></span>
<span><span class="va">island_status</span> <span class="op">&lt;-</span> <span class="fu"><a href="../reference/select_endemicity_status.html">select_endemicity_status</a></span><span class="op">(</span><span class="va">asr_geosse</span>, method <span class="op">=</span> <span class="st">"max"</span><span class="op">)</span></span>
<span></span>
<span><span class="co"># Add node data to the tree</span></span>
<span><span class="va">asr_geosse</span><span class="op">$</span><span class="va">island_status</span> <span class="op">&lt;-</span> <span class="va">island_status</span></span>
<span><span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">rownames</a></span><span class="op">(</span><span class="va">asr_geosse</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu">phylobase</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phylobase/man/nodeId-methods.html" class="external-link">nodeId</a></span><span class="op">(</span><span class="va">phylod</span>, <span class="st">"internal"</span><span class="op">)</span></span>
<span><span class="va">phylod</span> <span class="op">&lt;-</span> <span class="fu">phylobase</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phylobase/man/addData-methods.html" class="external-link">addData</a></span><span class="op">(</span></span>
<span>  <span class="va">phylod</span>,</span>
<span>  node.data <span class="op">=</span> <span class="va">asr_geosse</span></span>
<span><span class="op">)</span></span>
<span><span class="fu"><a href="../reference/plot_phylod.html">plot_phylod</a></span><span class="op">(</span><span class="va">phylod</span><span class="op">)</span></span></code></pre></div>
<p><img src="Extending_asr_files/figure-html/fit_geosse_and_asr-1.png" width="700"></p>
<p>We are now ready to extract the island clade.</p>
<div class="sourceCode" id="cb31"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">island_clades</span> <span class="op">&lt;-</span> <span class="fu">DAISIEprep</span><span class="fu">::</span><span class="fu"><a href="../reference/extract_island_species.html">extract_island_species</a></span><span class="op">(</span></span>
<span>  phylod <span class="op">=</span> <span class="va">phylod</span>,</span>
<span>  extraction_method <span class="op">=</span> <span class="st">"asr"</span></span>
<span><span class="op">)</span></span>
<span><span class="co">#&gt; Warning in extract_species_asr(phylod = phylod, species_label = as.character(phylod@label[i]), : Root of the phylogeny is on the island so the colonisation</span></span>
<span><span class="co">#&gt;               time from the stem age cannot be collected, colonisation time</span></span>
<span><span class="co">#&gt;               will be set to infinite.</span></span>
<span><span class="va">island_clades</span><span class="op">@</span><span class="va">island_tbl</span><span class="op">$</span><span class="va">species</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1] "bird_d" "bird_o" "bird_p" "bird_e" "bird_f" "bird_b" "bird_g" "bird_m"</span></span>
<span><span class="co">#&gt; [9] "bird_n"</span></span>
<span><span class="va">island_clades</span><span class="op">@</span><span class="va">island_tbl</span><span class="op">$</span><span class="va">branching_times</span></span>
<span><span class="co">#&gt; [[1]]</span></span>
<span><span class="co">#&gt; [1] 8.0047889 6.4518206 4.2518993 2.5503507 1.6635537 1.6299262 0.4224813</span></span>
<span><span class="co">#&gt; [8] 0.4212151</span></span></code></pre></div>
<p>In the case of this example, the crown node is inferred as present on
the island (nonendemic), such that the time of the initial colonisation
event cannot be estimated from the tree. Branching times are collected,
but when passing this data to DAISIE, colonisation time will by default
be set to <code>-Inf</code>, that is the age of the island.</p>
</div>
<div class="section level2">
<h2 id="corhmm">corHMM<a class="anchor" aria-label="anchor" href="#corhmm"></a>
</h2>
<p>This example is currently a work-in-progress, and will added to the
vignette soon.</p>
</div>
<div class="section level2">
<h2 id="concluding-remarks">Concluding remarks<a class="anchor" aria-label="anchor" href="#concluding-remarks"></a>
</h2>
<p>Extracting an island dataset from a larger phylogeny may in case
require estimation of the history of colonisation of the island within
the internal nodes of the phylogeny. We have shown how to do so above
using examples from three R packages commonly used for ancestral state
reconstruction. The choice of the most appropriate model to use for
ancestral state reconstruction is a difficult question, and an integral
part of the scientific exercise. This is likely to depend entirely on
both the taxa and island system at hand, and thus we believe the user is
best placed to make an informed decision. It is entirely possible to use
multiple models to assess the sensibility of island data extraction to
some processes, for example the possibility of jump dispersal (DEC+J) or
feedbacks between speciation and island status (GeoSSE). Similarly, one
may find the <code>method = "random"</code> option of
<code><a href="../reference/select_endemicity_status.html">select_endemicity_status()</a></code> of use to study the uncertainty
of reconstructions within the output of a single model. Through these
examples, we hope to have introduced an accessible framework, that can
be modified with ease to fit the specific needs of the analysis at
hand.</p>
</div>
<div class="section level2">
<h2 id="appendix">Appendix<a class="anchor" aria-label="anchor" href="#appendix"></a>
</h2>
<p>Below, we show the data manipulation underlying ASR with maximum
parsimony, as done inside <code><a href="../reference/add_asr_node_states.html">add_asr_node_states()</a></code>.</p>
<div class="sourceCode" id="cb32"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">tip_states</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="op">)</span></span>
<span><span class="va">endemicity_status</span> <span class="op">&lt;-</span> <span class="fu">phylobase</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phylobase/man/tdata-methods.html" class="external-link">tipData</a></span><span class="op">(</span><span class="va">phylod</span><span class="op">)</span><span class="op">$</span><span class="va">endemicity_status</span></span>
<span><span class="kw">for</span> <span class="op">(</span><span class="va">i</span> <span class="kw">in</span> <span class="fu"><a href="https://rdrr.io/r/base/seq.html" class="external-link">seq_along</a></span><span class="op">(</span><span class="va">endemicity_status</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"^not_present$"</span>, x <span class="op">=</span> <span class="va">endemicity_status</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">tip_states</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">1</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"^nonendemic$"</span>, x <span class="op">=</span> <span class="va">endemicity_status</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">tip_states</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">2</span></span>
<span>  <span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">grepl</a></span><span class="op">(</span>pattern <span class="op">=</span> <span class="st">"^endemic$"</span>, x <span class="op">=</span> <span class="va">endemicity_status</span><span class="op">[</span><span class="va">i</span><span class="op">]</span><span class="op">)</span><span class="op">)</span> <span class="op">{</span></span>
<span>    <span class="va">tip_states</span><span class="op">[</span><span class="va">i</span><span class="op">]</span> <span class="op">&lt;-</span> <span class="fl">3</span></span>
<span>  <span class="op">}</span></span>
<span><span class="op">}</span></span></code></pre></div>
<p>The maximum parsimony ancestral state reconstruction is from the R
package <code>castor</code>. The <code>castor</code> package only works
with S3 <code>phylo</code> objects, so we need to convert the phylogeny
back to this type and then run the analysis.</p>
<div class="sourceCode" id="cb33"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">phylo</span> <span class="op">&lt;-</span> <span class="fu">as</span><span class="op">(</span><span class="va">phylo</span>, <span class="st">"phylo"</span><span class="op">)</span></span>
<span><span class="va">asr</span> <span class="op">&lt;-</span> <span class="fu">castor</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/castor/man/asr_max_parsimony.html" class="external-link">asr_max_parsimony</a></span><span class="op">(</span></span>
<span>  tree <span class="op">=</span> <span class="va">phylo</span>,</span>
<span>  tip_states <span class="op">=</span> <span class="va">tip_states</span>,</span>
<span>  transition_costs <span class="op">=</span> <span class="st">"sequential"</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>This provides us with a matrix with the probabilities of the states
(island presence/absence) at each node in the phylogeny. The first
column of the matrix is not present on the island and the second column
of the matrix is present on the island.</p>
<div class="sourceCode" id="cb34"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">asr</span><span class="op">$</span><span class="va">ancestral_likelihoods</span><span class="op">)</span> <span class="op">==</span> <span class="fl">2</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">asr</span><span class="op">$</span><span class="va">ancestral_likelihoods</span><span class="op">)</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"not_present"</span>, <span class="st">"nonendemic"</span><span class="op">)</span></span>
<span><span class="op">}</span> <span class="kw">else</span> <span class="kw">if</span> <span class="op">(</span><span class="fu"><a href="https://rdrr.io/r/base/nrow.html" class="external-link">ncol</a></span><span class="op">(</span><span class="va">asr</span><span class="op">$</span><span class="va">ancestral_likelihoods</span><span class="op">)</span> <span class="op">==</span> <span class="fl">3</span><span class="op">)</span> <span class="op">{</span></span>
<span>  <span class="fu"><a href="https://rdrr.io/r/base/colnames.html" class="external-link">colnames</a></span><span class="op">(</span><span class="va">asr</span><span class="op">$</span><span class="va">ancestral_likelihoods</span><span class="op">)</span> <span class="op">&lt;-</span></span>
<span>    <span class="fu"><a href="https://rdrr.io/r/base/c.html" class="external-link">c</a></span><span class="op">(</span><span class="st">"not_present"</span>, <span class="st">"nonendemic"</span>, <span class="st">"endemic"</span><span class="op">)</span></span>
<span><span class="op">}</span></span>
<span><span class="va">asr</span><span class="op">$</span><span class="va">ancestral_likelihoods</span></span></code></pre></div>
<p>Once we have the matrix with the likelihood of the states we can
chose the most probable state at each node using the
<code><a href="https://rdrr.io/r/base/maxCol.html" class="external-link">max.col()</a></code> function. Here we need to make a decision that
will have downstream consequences for the DAISIE data extracted from the
tree, which is when a node has island presence and absence equally
probable we need to decide whether that species should be considered on
the island. To consider it on the island use
<code>ties.method = "last"</code> in the <code><a href="https://rdrr.io/r/base/maxCol.html" class="external-link">max.col()</a></code>
function, if you consider it not on the island use
<code>ties.method = "first"</code>. For this example we will assume that
species are on the island, but for completeness it may be worth running
both and then seeing if there are significant downstream
consequences.</p>
<div class="sourceCode" id="cb35"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">node_states</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/maxCol.html" class="external-link">max.col</a></span><span class="op">(</span><span class="va">asr</span><span class="op">$</span><span class="va">ancestral_likelihoods</span>, ties.method <span class="op">=</span> <span class="st">"last"</span><span class="op">)</span></span></code></pre></div>
<p>These values can now be converted back to string to make them more
readable.</p>
<div class="sourceCode" id="cb36"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">node_states</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span></span>
<span>  pattern <span class="op">=</span> <span class="st">"1"</span>, replacement <span class="op">=</span> <span class="st">"not_present"</span>, x <span class="op">=</span> <span class="va">node_states</span></span>
<span><span class="op">)</span></span>
<span><span class="va">node_states</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span></span>
<span>  pattern <span class="op">=</span> <span class="st">"2"</span>, replacement <span class="op">=</span> <span class="st">"nonendemic"</span>, x <span class="op">=</span> <span class="va">node_states</span></span>
<span><span class="op">)</span></span>
<span><span class="va">node_states</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/grep.html" class="external-link">gsub</a></span><span class="op">(</span></span>
<span>  pattern <span class="op">=</span> <span class="st">"3"</span>, replacement <span class="op">=</span> <span class="st">"endemic"</span>, x <span class="op">=</span> <span class="va">node_states</span></span>
<span><span class="op">)</span></span></code></pre></div>
<p>Now the ancestral states at the nodes is available we can combine it
into our <code>phylod</code> object.</p>
<div class="sourceCode" id="cb37"><pre class="downlit sourceCode r">
<code class="sourceCode R"><span><span class="va">node_data</span> <span class="op">&lt;-</span> <span class="fu"><a href="https://rdrr.io/r/base/data.frame.html" class="external-link">data.frame</a></span><span class="op">(</span></span>
<span>  island_status <span class="op">=</span> <span class="va">node_states</span>,</span>
<span>  endemic_prob <span class="op">=</span> <span class="va">asr</span><span class="op">$</span><span class="va">ancestral_likelihoods</span><span class="op">[</span>, <span class="st">"endemic"</span><span class="op">]</span>,</span>
<span>  nonendemic_prob <span class="op">=</span> <span class="va">asr</span><span class="op">$</span><span class="va">ancestral_likelihoods</span><span class="op">[</span>, <span class="st">"nonendemic"</span><span class="op">]</span>,</span>
<span>  not_present_prob <span class="op">=</span> <span class="va">asr</span><span class="op">$</span><span class="va">ancestral_likelihoods</span><span class="op">[</span>, <span class="st">"not_present"</span><span class="op">]</span>,</span>
<span>  row.names <span class="op">=</span> <span class="fu">phylobase</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phylobase/man/nodeId-methods.html" class="external-link">nodeId</a></span><span class="op">(</span><span class="va">phylod</span>, <span class="st">"internal"</span><span class="op">)</span></span>
<span><span class="op">)</span></span>
<span><span class="va">phylod</span> <span class="op">&lt;-</span> <span class="fu">phylobase</span><span class="fu">::</span><span class="fu"><a href="https://rdrr.io/pkg/phylobase/man/phylo4d-methods.html" class="external-link">phylo4d</a></span><span class="op">(</span></span>
<span>  <span class="va">phylo</span>,</span>
<span>  tip.data <span class="op">=</span> <span class="fu"><a href="https://rdrr.io/r/base/as.data.frame.html" class="external-link">as.data.frame</a></span><span class="op">(</span><span class="va">endemicity_status</span><span class="op">)</span>,</span>
<span>  node.data <span class="op">=</span> <span class="va">node_data</span></span>
<span><span class="op">)</span></span></code></pre></div>
</div>
<div class="section level2 unnumbered">
<h2 class="unnumbered" id="references">References<a class="anchor" aria-label="anchor" href="#references"></a>
</h2>
<div id="refs" class="references csl-bib-body hanging-indent" line-spacing="2">
<div id="ref-beaulieu_identifying_2013" class="csl-entry">
Beaulieu, J. M., B. C. O’Meara, and M. J. Donoghue. 2013. <a href="https://doi.org/10.1093/sysbio/syt034" class="external-link">Identifying <span>Hidden
Rate Changes</span> in the <span>Evolution</span> of a <span>Binary
Morphological Character</span>: <span>The Evolution</span> of
<span>Plant Habit</span> in <span>Campanulid Angiosperms</span></a>.
Systematic Biology 62:725–737.
</div>
<div id="ref-fitzjohn_diversitree_2012" class="csl-entry">
FitzJohn, R. G. 2012. <a href="https://doi.org/10.1111/j.2041-210X.2012.00234.x" class="external-link">Diversitree:
Comparative phylogenetic analyses of diversification in
<span>R</span></a>. Methods in Ecology and Evolution 3:1084–1092.
</div>
<div id="ref-goldberg_phylogenetic_2011" class="csl-entry">
Goldberg, E. E., L. T. Lancaster, and R. H. Ree. 2011. <a href="https://doi.org/10.1093/sysbio/syr046" class="external-link">Phylogenetic
<span>Inference</span> of <span>Reciprocal Effects</span> between
<span>Geographic Range Evolution</span> and
<span>Diversification</span></a>. Systematic Biology 60:451–465.
</div>
<div id="ref-louca_efficient_2018" class="csl-entry">
Louca, S., and M. Doebeli. 2018. <a href="https://doi.org/10.1093/bioinformatics/btx701" class="external-link">Efficient
comparative phylogenetics on large trees</a>. Bioinformatics
34:1053–1055.
</div>
<div id="ref-maddison_inferring_2006" class="csl-entry">
Maddison, W. P., and L. L. Knowles. 2006. <a href="https://doi.org/10.1080/10635150500354928" class="external-link">Inferring
<span>Phylogeny Despite Incomplete Lineage Sorting</span></a>.
Systematic Biology 55:21–30.
</div>
<div id="ref-matzke_probabilistic_2013" class="csl-entry">
Matzke, N. J. 2013. <a href="https://doi.org/10.21425/F5FBG19694" class="external-link">Probabilistic historical
biogeography: New models for founder-event speciation, imperfect
detection, and fossils allow improved accuracy and model-testing</a>.
Frontiers of Biogeography 5.
</div>
<div id="ref-ree_maximum_2008" class="csl-entry">
Ree, R. H., and S. A. Smith. 2008. <a href="https://doi.org/10.1080/10635150701883881" class="external-link">Maximum
<span>Likelihood Inference</span> of <span>Geographic Range
Evolution</span> by <span>Dispersal</span>, <span>Local
Extinction</span>, and <span>Cladogenesis</span></a>. Systematic Biology
57:4–14.
</div>
<div id="ref-valente_equilibrium_2015" class="csl-entry">
Valente, L. M., A. B. Phillimore, and R. S. Etienne. 2015. <a href="https://doi.org/10.1111/ele.12461" class="external-link">Equilibrium and non-equilibrium
dynamics simultaneously operate in the <span>Galápagos</span>
islands</a>. Ecology Letters 18:844–852.
</div>
</div>
</div>
  </main><aside class="col-md-3"><nav id="toc"><h2>On this page</h2>
    </nav></aside>
</div>



    <footer><div class="pkgdown-footer-left">
  <p></p>
<p>Developed by Joshua W. Lambert, Luis Valente, Lizzie Roeble, Theo Pannetier.</p>
</div>

<div class="pkgdown-footer-right">
  <p></p>
<p>Site built with <a href="https://pkgdown.r-lib.org/" class="external-link">pkgdown</a> 2.0.7.</p>
</div>

    </footer>
</div>

  

  

  </body>
</html>
